File: 14888.py
# 연산자 끼워넣기
# https://www.acmicpc.net/problem/14888

import sys
from itertools import permutations

input = sys.stdin.readline

def calc(a, operators):
    result = a[0]

    for i in range(0, len(operators)):
        if operators[i] == "+":
            result += a[i + 1]
        elif operators[i] == "-":
            result -= a[i + 1]
        elif operators[i] == "*":
            result *= a[i + 1]
        else:
            if result < 0:
                result = -(-result // a[i + 1])
            else:
                result //= a[i + 1]

    return result

n = int(input())
a = list(map(int, input().split()))

plus, minus, multiplication, division = map(int, input().split())
operators = []
operators.extend(["+"] * plus)
operators.extend(["-"] * minus)
operators.extend(["*"] * multiplication)
operators.extend(["/"] * division)

min_ans = 9876543210
max_ans = -9876543210
operator_permutations = set(permutations(operators, len(operators)))
for operators in operator_permutations:
    result = calc(a, operators)
    if result > max_ans:
        max_ans = result

    if result < min_ans:
        min_ans = result

print(max_ans)
print(min_ans)

File: 14888.kt
import java.util.*
import kotlin.math.max
import kotlin.math.min

// 연산자 끼워넣기
// https://www.acmicpc.net/problem/14888

lateinit var nums: IntArray
val operators = mutableListOf(0, 0, 0, 0)
const val PLUS = 0
const val MINUS = 1
const val MULTIPLICATION = 2
const val DIVISION = 3
var minNum = 1234567890
var maxNum = -1234567890

fun main() {
    input()
    solve()
    println(maxNum)
    println(minNum)
}

fun input() {
    val reader = Scanner(System.`in`)
    val n = reader.nextInt()
    nums = IntArray(n)

    for (i in 0 until n) {
        nums[i] = reader.nextInt()
    }

    for (operator in PLUS..DIVISION) {
        operators[operator] = reader.nextInt()
    }
}

fun solve() {
    calc(nums[0], 1)
}

fun calc(currentNum: Int, currentIndex: Int) {
    if (currentIndex == nums.size) {
        maxNum = max(maxNum, currentNum)
        minNum = min(minNum, currentNum)
    }

    for (operator in PLUS..DIVISION) {
        if (operators[operator] > 0) {
            val nextNum = when (operator) {
                PLUS -> {
                    currentNum + nums[currentIndex]
                }
                MINUS -> {
                    currentNum - nums[currentIndex]
                }
                MULTIPLICATION -> {
                    currentNum * nums[currentIndex]
                }
                else -> {
                    currentNum / nums[currentIndex]
                }
            }

            operators[operator] -= 1
            calc(nextNum, currentIndex + 1)
            operators[operator] += 1
        }
    }
}